### ISSUE TITLE
Get jailname for IP

### ISSUE THREAD

**Issue by srulikuk** [fail2ban/fail2ban#2725]
On my various servers I have about 100 different jails in total. 
The problem is when i need to unban an IP I need to go through all the jails to find which jail has this IP.

The question is is there a way to pass the IP to fail2ban-client and in response get the jailname OR is there a way to unban the IP without providing the jailname? 

Currently my to find the jailname I query the fail2ban's sqlite DB in my python script
```
SELECT jail
FROM bans
WHERE ip='{}'
""".format(
    ip
)
```
but I am not happy with this as many times it fails with an DB locked/busy error

Using fail2ban versions 0.10.2 to 0.10.5

**Comment by sebres** [comment]
> Get jailname for IP

Currently there is no such functionality available, but ...

> The problem is when i need to unban an IP I need to go through all the jails to find which jail has this IP.

No, you don't need (at least fail2ban >= 0.10), see https://github.com/fail2ban/fail2ban/issues/2349#issuecomment-465529331
And read [man of fail2ban-client](https://manpages.debian.org/buster/fail2ban/fail2ban-client.1.en.html).
So you don't need to know the jail to unban IPs in all jails.

**Comment by srulikuk** [comment]
Arrrgh, I don't know why i was under the impression that the jailname must be provided.
This solves it `fail2ban-client unban x.x.x.x`

Many thanks
(btw, is there a python function to unban ip? currently I am using a subprocess in my scripts to execute fail2ban-client commands)

**Comment by srulikuk** [comment]
and similarly is there a way to ignoreip in all jails? currently i use `fail2ban-client set jailname addignoreip x.x.x.x`

**Comment by sebres** [comment]
My pleasure.

> is there a python function to unban ip?

Sure, but I guess server-side API would be not really suitable for you (since your python part is probably not running in fail2ban-server instance), and client-side API is not really public...
But fail2ban communication is simple and based on a pickle serialization. so you could send a simple pickle packet to the server (without a fail2ban API at all):
```python
s = socket.socket(socket.AF_UNIX, socket.SOCK_STREAM)
s.connect('/run/fail2ban/fail2ban.sock')
s.send(pickle.dumps(['unban', '192.0.2.1'], 0) + b"<F2B_END_COMMAND>")
s.close()
```
This way you can use almost every command available in `fail2ban-client`, for example to notify fail2ban about a single failed attempt of some jail - https://github.com/fail2ban/fail2ban/issues/2559#issuecomment-548851675.

> is there a way to ignoreip in all jails? currently i use fail2ban-client set jailname addignoreip x.x.x.x

Not really. But you could try `ignorecommand` instead of `ignoreip`... it is fast enough if #2176 is enabled.
(Or if you have own DNS (or it is something like a dyndns IP), you can specify fqdn like `ignoreip = 127.0.0.1 ... fail2ban-ignore-ips-subnet.example.com` and simply extend/set it dynamically in DNS service).
Anyway better is to avoid recognizing of legitimate attempts as failures, so perhaps you can change you filter logic (RE's, increase maxretry, etc) instead of do ignore by IP.

**Comment by srulikuk** [comment]
Wow many thanks, 
`s.send(pickle.dumps(['unban', '192.0.2.1'], 0) + b"<F2B_END_COMMAND>")`
help me unpickle this please, what's the 0 "0)" and what's the b  "+ b" ?

Also would this work for ignoreip
`s.send(pickle.dumps(['set jailName addignoreip', '192.0.2.1'], 0) + b"<F2B_END_COMMAND>")`

With regards to ignore ip, it's not for self, I have a shared jail between all my servers, (using https://github.com/srulikuk/centralised-f2b - wip) and need to unban ~5 different customer IP's daily, (I unban it on any client using the script in centralised-f2b and that unbans it on all clients) some of those IP's I need to whitelist, those I add it to a whitelist-file but if I understand correctly the whitelist file is only read on reload so need a way to add these IP's to fail2ban until such time that is reloads.

In all jails I have.
```
[INCLUDES]
before = whitelist.local
ignoreip = %(ignoreip)s
```

**Comment by sebres** [comment]
> help me unpickle this please, what's the 0 "0)" and what's the b "+ b" ?

for 0, see `pickel.dumps` [docu](https://docs.python.org/2.7/library/pickle.html#usage) - this is a protocol (to be sure same version is used as in server regardless what your python prefers).

as for `b"<F2B_END_COMMAND>"` this is a mark of f2b protocol, just signaling end of command.
and you can send multiple commands (and to fulfill the proto - `b"<F2B_CLOSE_COMMAND>"` as end of stream).

As for ignore command, it is wrong, rather it'd be a list where every word is separate element like that:
```diff
- s.send(pickle.dumps(['set jailName addignoreip', '192.0.2.1'], 0) + b"<F2B_END_COMMAND>")
+ s.send(pickle.dumps(['set', jail, 'addignoreip', '192.0.2.1'], 0) + b"<F2B_END_COMMAND>")
```
And probably don't forget to remove "old" obsolete IPs.

> need to unban ~5 different customer IP's daily,

Looks like pretty well argument for dyndns similar handling (fixed FQDN with dynamic IPs) as I suggested above:
```ini
ignoreip = customer1.tld ... customer5.tld
```

**Comment by srulikuk** [comment]
again, many thanks.
with regards to ignoreip, most are not fixed locations, its users getting banned on 4G or home network or some other temp network, etc... those are only unbanned, I only need to whitelist about 1% of the ip's I unban, to put it in perspective I have about 200k banned IP's, in total I had to unban ~200, of those 3 made it into my whitelist.

We opened a jar of pickles here ;-)
How would I use pickle to get a list of current jails (the output of fail2ban-client status)

**Comment by sebres** [comment]
Firstly, thank you very much for the sponsorship! 🙇‍♂️ 🎉 
This is really... unexpected.... very often the single feedback, that one gets, is "why your f**ng program doesn't work as I want" or "why you'd closing my issue which is not yet solved" 😄

> How would I use pickle to get a list of current jails (the output of fail2ban-client status)

Well, because you have to receive a data from server too, it would be indeed better to use some client API ([class fail2ban.client.csocket.CSocket](https://github.com/fail2ban/fail2ban/blob/0.10/fail2ban/client/csocket.py)) instead of pickle serialization directly (if you want it nevertheless see how `CSocket.send` work)...

This will do the job (it returns a comma-separated string with jail names):
```bash
$ fail2ban-python -c 'from fail2ban.client.csocket import CSocket; s = CSocket("/run/fail2ban/fail2ban.sock"); print(s.send(["status"])[1][1][1]); s.close()'
jail1, jail2, ..., jailN
```
Bad is here, that the real output of a status command is prepared for the beautifying mechanism of the fail2ban-client, so at the moment it is a tuple in form of
`(0, [("Number of jail", len(jails)), ("Jail list", jailList)])`
where `jailList` is a string `", ".join(jails)`, so to get this value you have to use `[1][1][1]` and following split to make a list again.
So something like that will work now:
```python
from fail2ban.client.csocket import CSocket
s = CSocket("/run/fail2ban/fail2ban.sock")
jails = s.send(["status"])[1][1][1]
s.close()
jails = jails.split(", ")
for j in jails:
   print(j)
```
May be we need to extend F2B for a raw status (or command returning raw jail-list as list).
Or just some command (like `raw on`) that will signal the server, this connection prefers raw output rather.

Also note that however we're trying to retain backwards compatibility by API also, it can happen that some things (e. g. protocol) may be changed in some version (due to bug fixing, simplification or some other still unknown reasons).

**Comment by srulikuk** [comment]
thanks again ;-)

I will use this array to add ignoreip to each jail, can I also use fail2ban.client.csocket to unban an IP?

With regards to changes in future versions, this is what worries me most but I really prefer this to using subprocess.

p.s A little cheeky and way off topic, do you know if there as a way in python to check ipset for an ip (in bash: ipset list | grep "x.x.x.x") 

(You have taken your time to answer my question and i really appreciate it, I have sponsored as much as I could at the moment even though it doest come close to the time you spent helping me on my issues. (OTHERS PLEASE TAKE NOTE - The people maintaing and supporting openSource projects also need to eat!!!!!)

**Comment by sebres** [comment]
> can I also use fail2ban.client.csocket to unban an IP?

Sure, you can. `.send(["unban", ip])`

> or to check if an IP is banned?

To unban it, you don't need to check it is banned (it will do nothing if IP is not found in any jail), but there is no command to check IP is banned at the moment, excepting probably to take a look in output of status for every jail (using `['status', jail]`).
So right now probably a look into database is simplest solution here (see https://github.com/fail2ban/fail2ban/issues/2659#issuecomment-599271261 for similar question, and extend it with `and ip = ?` criteria).

> it doesn't come close to the time you spent helping me on my issues

Time is indeed the lacking resource mostly (and it is not my single project where I'm contributing), but I'm happy to help where I can also for free. But thank you again.

> people maintaing and supporting openSource projects also need to eat

"Be nice to your developer" would be often enough too. 😄

**Comment by srulikuk** [comment]
I am trying to get away from querying the sqllite DB as I am getting too many DB locked/busy errors.
Currently I am using subprocess to query "ipset list" for the IP, this is way of topic but as I want to move away from using subprocess do you know if there is a python way to query ipset?

**Comment by srulikuk** [comment]
> Sure, you can. `.send(["unban", ip])`
This works fine however I cannot figure out how to use this for addignoreip or for reload fail2ban-client, i tried various different ways, when getting the output for addignoreip it seems it does not know the command and for reload it wants the jailname and some option, not sure which option.

Any ideas?
Many thanks

**Comment by sebres** [comment]
> I want to move away from using subprocess do you know if there is a python way to query ipset?

Not that I know... IMHO, ipsetpy also uses subprocess to manage ipsets.

We can just provide this functionality via fail2ban-client/protocol (something like `test <IP> ?<JAIL>?`). It will be very similar `unban <IP>`. I'll try to provide it later.

>  i tried various different ways, when getting the output for addignoreip it seems it does not know the command 

Hmmm... What do you mean? Does this one not work for you?
```
from fail2ban.client.csocket import CSocket
s = CSocket("/run/fail2ban/fail2ban.sock")
jails = s.send(['set', 'sshd', 'addignoreip', '192.0.2.1'])
s.close()
```
It works on my side pretty well:
```bash
# python -c 'from fail2ban.client.csocket import CSocket; s = CSocket("/run/fail2ban/fail2ban.sock"); print(s.send(["set", "sshd", "addignoreip", "192.0.2.1"])); s.close()'
(0, ['127.0.0.0/8', 'mysubnet.tld', '192.0.2.1'])
# python -c 'from fail2ban.client.csocket import CSocket; s = CSocket("/run/fail2ban/fail2ban.sock"); print(s.send(["set", "sshd", "delignoreip", "192.0.2.1"])); s.close()'
(0, ['127.0.0.0/8', 'mysubnet.tld'])
```
Normally all the commands you would see in following dump are possible directly over a socket (f2b protocol):
```
fail2ban-client --dp
```
> for reload it wants the jailname and some option, not sure which option.

A reload via protocol directly is a bit complex (you have to supply several commands from above).

**Comment by srulikuk** [comment]
> Not that I know... IMHO, ipsetpy also uses subprocess to manage ipsets.

Yes - I checked the lib it uses subproccess

> We can just provide this functionality via fail2ban-client/protocol (something like `test <IP> ?<JAIL>?`). It will be very similar `unban <IP>`. I'll try to provide it later.

Great

> jails = s.send(['set', 'sshd', 'addignoreip', '192.0.2.1'])

Works now, I must have done something wrong before, not sure what.
> Normally all the commands you would see in following dump are possible directly over a socket (f2b protocol):
> ```
> fail2ban-client --dp
> ```

This is great, will use as partial man page.

**Comment by sebres** [comment]
Banned info is implemented now in 54b2208690e3c2fff00fbd9b197984d880e29a02.
It can provide that in several ways:
- `['banned']` returns dictionary in form of `{'jail1', [list of banned IPs], ...}`
- `['banned', 'IP1', ..., 'IPn']` returns list with list of jails where given IPs are banned, like `[['jail1', 'jail2'], ..., ['jail=of-IP-N']]`
- `['get', jail, 'banned']` returns list of banned IPs of jail in form of `[list of banned IPs]`
- `['get', jail, 'banned', ip]` returns 1 if given single IP is banned in given jail or 0 elsewhere
- `['get', jail, 'banned', 'IP1', ..., 'IPn']` returns a list of banned in given jail values (1 or 0) corresponding every given IP, like `[1, ..., 1]`.

**Comment by srulikuk** [comment]
thanks you so much, how would i patch my distro's version?

**Comment by sebres** [comment]
well, basically you can try to patch it... but proper way would be to install it from source (branch 0.10-extend-proto-banned) like described in readme or to wait unless it going released for your distribution.

**Comment by sebres** [comment]
The new commands are available in 0.10 since b8e2b772656e488be3abebcfff622c797e1f664b, and also merged in all major branches (0.11, 1.0).
Have you accomplished to patch or to test that in the time?

**Comment by srulikuk** [comment]
@sebres 
I am very sorry for the long delay in responding, I had some unexpected stuff to deal with.
With regards to testing the updates, I need my scripts to work on distro fail2ban versions, as of today I checked the fail2ban in 2 distros I use and both do not have these options yet, will have to use subprocess in the meanwhile.
Once again sorry for the delay and many thanks for your help
